### ISSUE TITLE
Fails to detect some cases of unavailable upstream index

### ISSUE THREAD

**Issue by hchargois** [devpi/devpi#549]
I want to use devpi to allow resilient deployments, i.e. still allow to run `pip install -r requirements.txt` even when PyPI is down.

I have tested devpi with the default configuration which should allow this, and tried to simulate PyPI being down. Depending on how I do such simulation, devpi does or doesn't serve its local packages. Here are the different ways I tried.

## Test setup

Using devpi-server 4.5.0, out of the box config.

First, I installed a package normally with pip to load it in the devpi cache, everything works.

In all the following cases, I edited my `/etc/hosts` to add the line:

    127.0.0.2 pypi.org

## Simulate a connection refused: works

With nothing listening on `127.0.0.2:443`, a connection to our `pypi.org` will be refused:

    $ curl https://pypi.org
    curl: (7) Failed to connect to pypi.org port 443: Connection refused

This scenario works with devpi. It attemps to connect, sees that it fails and serves the local package instead.

## Simulate a connection timeout: fails

Let's simulate a connection timeout by dropping packets to our fake server:

    $ sudo iptables -A INPUT -p tcp -d 127.0.0.2 --dport 443 -j DROP
    $ curl https://pypi.org
    curl: (7) Failed to connect to pypi.org port 443: Connection timed out

Now devpi doesn't want to serve the local package and pip retries and ultimately fails. See attached logs.

## Simulate a read timeout: fails

Now lets remove our iptables rule and put a basic server to accept connections but never write any response to them:

    $ sudo iptables -D INPUT -p tcp -d 127.0.0.2 --dport 443 -j DROP
    $ sudo ncat -l -k 127.0.0.2 443
    [...]
    $ curl https://pypi.org
    curl: (28) Operation timed out after 300541 milliseconds with 0 out of 0 bytes received

Basically same thing here, devpi timeouts but doesn't serve the local package, and ultimately fails. See attached logs.

## Others?

I didn't try other cases, especially none with an actual connection to a webserver. The webserver could reply with an unexpected status code, reset the connection, or it could be the TLS handshake failing, so many things can go wrong.

[pip-connection-timeout.log](https://github.com/devpi/devpi/files/2098492/pip-connection-timeout.log)
[devpi-connection-timeout.log](https://github.com/devpi/devpi/files/2098493/devpi-connection-timeout.log)
[devpi-read-timeout.log](https://github.com/devpi/devpi/files/2098490/devpi-read-timeout.log)
[pip-read-timeout.log](https://github.com/devpi/devpi/files/2098491/pip-read-timeout.log)

**Comment by fschulze** [comment]
For the timeout cases, there was a PR merged recently, but it's not released yet:
https://github.com/devpi/devpi/pull/544
Could you try that with a shorter timeout? I guess 5 seconds would be a much saner default timeout.

**Comment by fschulze** [comment]
Wrong status codes should be covered by the current code base, TLS handshake failures should be caught as well, not sure about connections reset, but since the TCP connection will break down in that case, I think that should work as well.
The biggest issue here is to simulate and test all these failure cases, especially for automated tests.

**Comment by hchargois** [comment]
Using a 5-second timeout makes the two timeout cases work. Devpi tries to connect/read a response from the upstream for 5-seconds, then timeouts, then serves the local package; pip is happy.

**Comment by fschulze** [comment]
Fix released with devpi-server 4.6.0.
